compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs

```diff
@@ -479,49 +516,106 @@ generateSrcFromLf thisPkgId pkgMap m = noLoc mod
         noLoc .
         mkOrig (mkModule damlStdlibUnitId $ mkModuleName "DA.Internal.Prelude") .
         mkOccName varName
-    mkTyConType :: TyCon -> HsType GhcPs
-    mkTyConType tyCon =
-        let name = getName tyCon
-         in HsTyVar NoExt NotPromoted . noLoc $
-            mkOrig (nameModule name) (occName name)
-    mkGhcPrimImport :: Bool -> String -> LImportDecl GhcPs
-    mkGhcPrimImport qualified  modName =  noLoc $
-           ImportDecl
-               { ideclExt = NoExt
-               , ideclSourceSrc = NoSourceText
-               , ideclName = noLoc $ mkModuleName modName
-               , ideclPkgQual =
-                     Just $ StringLiteral NoSourceText $ mkFastString "daml-prim"
-               , ideclSource = False
-               , ideclSafe = False
-               , ideclImplicit = False
-               , ideclQualified = qualified
-               , ideclAs = Nothing
-               , ideclHiding = Nothing
-               }
-    imports =
-        -- first, imports that we need in any case
-        map (mkGhcPrimImport True) ["GHC.Types", "GHC.Err"] ++
-        -- qualified imports from daml-prim
-        map (mkGhcPrimImport False) ["Data.String"] ++
-        -- unqualified importts from daml-prim
+    mkTyConType = mkTyConType' dontQualify
+    mkTyConTypeUnqual = mkTyConType' True
+    mkTyConType' :: Bool -> TyCon -> HsType GhcPs
+    mkTyConType' dontQualify tyCon
+        | dontQualify = HsTyVar noExt NotPromoted . noLoc $ mkRdrUnqual (occName name)
+        | otherwise =
+            HsTyVar noExt NotPromoted . noLoc $
+            mkRdrQual (moduleName $ nameModule name) (occName name)
+      where
+        name = getName tyCon
+    imports = declImports ++ additionalImports
+    additionalImports =
+        [ noLoc $
+        ImportDecl
```

Comment:
It might be helpful to factor this into a helper function to make it easy to see which fields of the import decl are different between different imports and to capture the pattern of filtering out the import on the module itself.

Reply:
I agree.